app/daemon.go

```diff
@@ -132,17 +142,28 @@ func startManager(c *cli.Context) error {
 		return fmt.Errorf("BUG: failed to detect the node IP")
 	}
 
-	done := make(chan struct{})
+	ctx := signals.SetupSignalContext()
```

Comment:
Can you describe a little bit more about this kind of context?

Reply:
longhorn-manager made a channel `done` and registered it as a shutdown channel to receive shut down signal by `util.RegisterShutdownChannel(done)`
it then passed it to the controllers and the manager API service

For webhook/recovery_backend, it created a context by `ctx := signals.SetupSignalContext()`
then passed it to the API service 

both actually have the same behavior and the purpose is to receive the shutdown signal.
I don't want to make a channel for manager API and context for webhook/recovery_backend.

So I decide to make a context only, and pass ctx.Done as the done channel used in manager API.

Reply:
- longhorn-manager registered the channel to receive the signal: 
https://github.com/longhorn/longhorn-manager/blob/master/app/daemon.go#L203-L207
- RegisterShutdownChannel
https://github.com/longhorn/longhorn-manager/blob/master/util/util.go#L348-L356

- Same as SetupSignalContext (`vendor/github.com/rancher/wrangler/pkg/signals/signal.go`)
```
// SetupSignalContext is same as SetupSignalHandler, but a context.Context is returned.
// Only one of SetupSignalContext and SetupSignalHandler should be called, and only can
// be called once.
func SetupSignalContext() context.Context {
	close(onlyOneSignalHandler) // panics when called twice

	shutdownHandler = make(chan os.Signal, 2)

	ctx, cancel := context.WithCancel(context.Background())
	signal.Notify(shutdownHandler, shutdownSignals...)
	go func() {
		<-shutdownHandler
		cancel()
		<-shutdownHandler
		os.Exit(1) // second signal. Exit directly.
	}()

	return ctx
}
```

Reply:
Oh Sorry, this is already used by webhook and recovery_backend. 
Just checked this context. It registers SIGTERM and SIGINT as channel stop signals. This is what we need here.

But here another question is, could a crash of one service (conversion webhook, validation webhook, etc ) lead to channel closing or other services stop?

Reply:
The root **context** is created in `daemon.go` and is passed down to manager/webhook/recovery these services
**cancel** is not passed down to services so no one can close the channel.
It only `cancel()` when the shutdown signal is received in the above`SetupSignalContext()` function.

Reply:
Hi @shuo-wu 
you might be right, the I only run http server in goroutine
```
go s.ListenAndServe()
```
but the goroutine failed won't lead to other services stopped which is not expected.

Let me fix it.

Reply:
I have fixed it by capturing the error returning from ListenAndServe() and fatal exit the program
WDYT?

Reply:
The fatal exit will lead to `ctx.Done()` close then all services/the whole pod stop, right? Then that LGTM

Reply:
it won't lead to ctx.Done()
but the main thread would log fatal and exit(1) and the whole pod will restart.